scripts/validations.py

```diff
@@ -0,0 +1,142 @@
+import itertools
+import json
+import sys
+
+ACTION_CONTEXT = 'context'
+ACTION_COUNT = 'count'
+ACTIONS = (ACTION_CONTEXT, ACTION_COUNT)
+
+
+def parse_validations(results):
+    return (json.loads(result) for result in results)
+
+
+def unlisted_validations(results, unlisted_addons=None):
+    if unlisted_addons is None:
+        unlisted_addons = get_unlisted_addons()
+    return (result
+            for result in results
+            if ('id' in result['metadata'] and
+                (not result['metadata'].get('listed', True)
+                 or result['metadata']['id'] in unlisted_addons)))
+
+
+def severe_validations(results):
+    return (result
+            for result in results
+            if (result['signing_summary']['high'] > 0 or
+                result['signing_summary']['medium'] > 0 or
+                result['signing_summary']['low'] > 0))
+
+
+def error_messages(results):
+    return ({'addon': result['metadata']['id'],
+             'message_id': '.'.join(message['id']),
+             'context': message['context']}
+            for result in results
+            for message in result['messages']
+            if 'signing_severity' in message)
+
+
+def sort_by_message(results):
+    return sorted(results, key=lambda r: r['message_id'])
+
+
+def group_by_message(results):
+    return itertools.groupby(results, lambda r: r['message_id'])
+
+
+def extract_error_results(results):
+    for error, messages in results:
+        all_messages = list(messages)
+        yield {
+            'error': error,
+            'total': len(all_messages),
+            'unique': len(set(msg['addon'] for msg in all_messages)),
+            'contexts': [msg['context'] for msg in all_messages],
+        }
+
+
+def sort_results_by_unique(results):
+    return sorted(results, reverse=True, key=lambda r: r['unique'])
+
+
+def format_error_count(results):
+    return ('{error} {total} {unique}'.format(**result)
+            for result in results)
+
+
+def format_contexts(results):
+    for result in results:
+        for context in result['contexts']:
+            yield json.dumps({
+                'error': result['error'],
+                'context': context,
+            })
+
+
+def memoize(fn):
+    """Memoize a 0-arg function."""
+    # We're going to store the result in a list so we can mutate it.
+    result = []
```

Comment:
Python 2 doesn't have `nonlocal` and `global` only works with globals. So you can't assign to `result`. You can mutate an object though so you can append it to a list and check the length as I do here. Another benefit is that this will cache a null or falsey value as well if that's a possibility.

I think you could define result on `fn` and it would work but that feels dirty. Using a sentinel that isn't `None` is probably best.

```py
def memoize(fn):
    fn._cached_result = sentinel = type('sentinel', (), {}) 
    def inner():
        if fn._cached_result is sentinel:
            fn._cached_result = fn()
        return fn._cached_result
    return inner